src/dune_pkg/rev_store.ml

```diff
@@ -7,20 +7,61 @@ module Re = Dune_re
 module Flock = Dune_util.Flock
 open Fiber.O
 
-type t = { dir : Path.t }
+module Object = struct
+  type t = Sha1 of string
+
+  let compare (Sha1 x) (Sha1 y) = String.compare x y
+  let to_string (Sha1 s) = s
+  let equal (Sha1 x) (Sha1 y) = String.equal x y
+  let to_dyn (Sha1 s) = Dyn.string s
+  let hash (Sha1 s) = String.hash s
+
+  type resolved = t
+
+  let of_sha1 s =
+    if String.length s = 40
+       && String.for_all s ~f:(function
+         | '0' .. '9' | 'a' .. 'f' | 'A' .. 'F' -> true
```

Comment:
Yes, but to disambiguate the object from a ref would require always querying the remote. Which means that it would be impossible to do a build without listing all the refs on all the remotes. So I think pretending that every such 40 char string is a sha1 is the lesser evil.

Reply:
Yes, I run into this exact issue before, which is why I was considering asking the user to specify the reference not in the ambiguous `#hash` format but rather specify whether it was a revision or a reference. OPAM doesn't have this issue as they can always refer to the remote. Maybe extending the repo specification format to specify `branch`/`tag` etc. would be the less fragile option than trying to guess.

Reply:
That would be a good option as well. It was on my TODO list, but it's low priority so I doubt we'll get around to it in time. If you'd like, you can write a proposal for a different format for source specifications. I would love to have a shorter form for github sources as well.